File: CS/Security.Extensions/RestorePasswordParameters.cs
using System;
using DevExpress.Xpo;
using DevExpress.ExpressApp;
using DevExpress.Data.Filtering;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp.Model;
using System.Security.Cryptography;
using DevExpress.Persistent.Validation;
using DevExpress.Persistent.Base.Security;
using DevExpress.ExpressApp.DC;

namespace Security.Extensions {
    [DomainComponent]
    [ModelDefault("Caption", "Restore Password")]
    [ImageName("Action_ResetPassword")]
    public class RestorePasswordParameters : ILogonActionParameters {
        public const string ValidationContext = "RestorePasswordContext";
        [RuleRequiredField(null, ValidationContext)]
        [RuleRegularExpression(null, ValidationContext, ManageUsersOnLogonController.EmailPattern)]
        public string Email { get; set; }
        public void Process(IObjectSpace objectSpace) {
            if (string.IsNullOrEmpty(Email))
                throw new ArgumentException("Email address is not specified!");
            IAuthenticationStandardUser user = objectSpace.FindObject(SecurityExtensionsModule.SecuritySystemUserType, CriteriaOperator.Parse("Email = ?", Email)) as IAuthenticationStandardUser;
            if (user == null)
                throw new ArgumentException("Cannot find registered users by the provided email address!");
            byte[] randomBytes = new byte[6];
            new RNGCryptoServiceProvider().GetBytes(randomBytes);
            string password = Convert.ToBase64String(randomBytes);
            //Dennis: Resets the old password and generates a new one.
            user.SetPassword(password);
            user.ChangePasswordOnFirstLogon = true;
            objectSpace.CommitChanges();
            EmailLoginInformation(Email, password);
        }
        public static void EmailLoginInformation(string email, string password) {
            //Dennis: Send an email with the login details here. 
            //Refer to http://msdn.microsoft.com/en-us/library/system.net.mail.mailmessage.aspx for more details.
            //throw new UserFriendlyException("Password recovery link was sent to " + email);
        }
    }
}

File: VB/Security.Extensions/RestorePasswordParameters.vb
Imports System
Imports DevExpress.Xpo
Imports DevExpress.ExpressApp
Imports DevExpress.Data.Filtering
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Model
Imports System.Security.Cryptography
Imports DevExpress.Persistent.Validation
Imports DevExpress.Persistent.Base.Security
Imports DevExpress.ExpressApp.DC

Namespace Security.Extensions
	<DomainComponent, ModelDefault("Caption", "Restore Password"), ImageName("Action_ResetPassword")>
	Public Class RestorePasswordParameters
		Implements ILogonActionParameters

		Public Const ValidationContext As String = "RestorePasswordContext"
		<RuleRequiredField(Nothing, ValidationContext), RuleRegularExpression(Nothing, ValidationContext, ManageUsersOnLogonController.EmailPattern)>
		Public Property Email() As String
		Public Sub Process(ByVal objectSpace As IObjectSpace) Implements ILogonActionParameters.Process
			If String.IsNullOrEmpty(Email) Then
				Throw New ArgumentException("Email address is not specified!")
			End If
			Dim user As IAuthenticationStandardUser = TryCast(objectSpace.FindObject(SecurityExtensionsModule.SecuritySystemUserType, CriteriaOperator.Parse("Email = ?", Email)), IAuthenticationStandardUser)
			If user Is Nothing Then
				Throw New ArgumentException("Cannot find registered users by the provided email address!")
			End If
			Dim randomBytes(5) As Byte
			CType(New RNGCryptoServiceProvider(), RNGCryptoServiceProvider).GetBytes(randomBytes)
			Dim password As String = Convert.ToBase64String(randomBytes)
			'Dennis: Resets the old password and generates a new one.
			user.SetPassword(password)
			user.ChangePasswordOnFirstLogon = True
			objectSpace.CommitChanges()
			EmailLoginInformation(Email, password)
		End Sub
		Public Shared Sub EmailLoginInformation(ByVal email As String, ByVal password As String)
			'Dennis: Send an email with the login details here. 
			'Refer to http://msdn.microsoft.com/en-us/library/system.net.mail.mailmessage.aspx for more details.
			'throw new UserFriendlyException("Password recovery link was sent to " + email);
		End Sub
	End Class
End Namespace
